### ISSUE TITLE
Projections can be made classes but you can't define instances

### ISSUE THREAD

**Issue by gmalecha** [coq/coq#8994]
#### Version

```shell
$ coqtop -v
The Coq Proof Assistant, version 8.8.2 (November 2018)
compiled on Nov 7 2018 21:54:20 with OCaml 4.07.0
```

#### Operating system

Ubuntu 18.04

#### Description of the problem

```coq
Set Primitive Projections.

Class type (t : Type) : Type :=
  { bar : t -> Prop  }.

Instance type_nat : type nat :=
{ bar := fun _ => True }.

Existing Class bar.

Global Instance foo_bar {n : nat} : bar n := I.
(* with primitive projections, this line fails with "Error: bar is not a declared type class."
 * without primitive projections it succeeds
 *)
```

Expected behavior: `bar` should be a typeclass so there shouldn't be a problem with the definition of `foo_bar` as an instance.
If there is a reason to forbid supporting this, then there should be an error on the call to `Existing Class bar`.

**Comment by ppedrot** [comment]
The problem here is that `bar n` in interpreted as the primitive projection `n.(bar)`. To fix this in a future-proof way, we need to also natively handle projections in typeclasse instances. There might be a day where we don't declare the compatibility projection constants.

**Comment by gmalecha** [comment]
Are you saying that `Existing Class bar` is making the compatibility *definition* a class, but not the projection itself?

**Comment by JasonGross** [comment]
> Are you saying that `Existing Class bar` is making the compatibility _definition_ a class, but not the projection itself?

Yes, it is exactly this.  Here is a workaround that @ppedrot will probably say I'd be terrible for relying on:
```coq
Set Primitive Projections.
Class type (t : Type) : Type :=
  { bar : t -> Prop  }.

Instance type_nat : type nat :=
  { bar := fun _ => True }.

Existing Class bar.

Global Instance foo_bar {n : nat} : match @bar with x => x _ _ n end := I.
```

**Comment by JasonGross** [comment]
However, note that this workaround, while it gets the statement to be accepted, probably does not solve your problem.  The instance gets picked up on
```coq
Check _ : match @bar with x => x _ _ _ end.
```
but not on
```coq
Check _ : @bar _ _ _.
```

**Comment by gmalecha** [comment]
Thanks, @JasonGross . I definitely don't want to do the workaround :-). This isn't a crucial issue for me right now.

**Comment by Alizter** [comment]
@ppedrot Isn't this closed by #9711 as well?

**Comment by ppedrot** [comment]
Indeed. We should add a test before closing this issue though.
